### PR TITLE
Add configuration option to run code analysis

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add "/p:Analysis=True" to the msbuild command to run code analysis. Separate PR to add it to CI/CD pipeline to follow.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

**Review by shankarseal — APPROVED** [review]

**Review by dthaler — COMMENTED** [review]
I notice the vcxproj files under tests/ and tools/ are not yet included

**Comment by dthaler** [comment]
> @dthaler Why would we run analysis over test code? That seems redundant.

To catch bugs in tests.  Code analysis is applicable to everything.

**Comment by Alan-Jowett** [comment]
To make things simpler, I am moving the common code analysis settings to a .props file and including it from all the projects. As we update the analysis setting, all projects will pick it up.

**Comment by Alan-Jowett** [comment]
Enabled all analysis rules that currently pass; document and suppress rules that trigger failures.

File: Analyze.default.props [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: MIT
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" TreatAsLocalProperty="Platform">
+  <PropertyGroup Condition="'$(Analysis)'=='True'">
+    <RunCodeAnalysis>true</RunCodeAnalysis>
+    <DisableAnalyzeExternal>true</DisableAnalyzeExternal>
+    <CodeAnalysisRuleSet>$(SolutionDir)\Analyze.default.ruleset</CodeAnalysisRuleSet>
```
Comment (dthaler) [inline Analyze.default.props:10]:
```suggestion
    <CodeAnalysisRuleSet>$(SolutionDir)Analyze.default.ruleset</CodeAnalysisRuleSet>
```
SolutionDir always ends in a \ already

File: Analyze.default.ruleset [hunk 2]
```diff
@@ -0,0 +1,281 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: MIT
+-->
+<RuleSet Name="Copy of Microsoft Native Recommended Rules" Description="These rules focus on the most critical and common problems in your native code, including potential security holes and application crashes.  You should include this rule set in any custom rule set you create for your native projects.  This ruleset is designed to work with Visual Studio Professional edition and higher." ToolsVersion="16.0">
+  <Rules AnalyzerId="Microsoft.Analyzers.NativeCodeAnalysis" RuleNamespace="Microsoft.Rules.Native">
+    <!-- All rules marked as Warning should be switched to Error once all instances are resolved. -->
+    <!-- Resolving an instance via either locally suppressing it or fixing the issue. -->
```
Comment (dthaler) [inline Analyze.default.ruleset:9]:
```suggestion
    <!-- Resolving an instance can be done via either locally suppressing it or fixing the issue. -->
```
Couldn't parse grammar.  Is above what was intended?

File: Analyze.default.ruleset [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    <Rule Id="C6331" Action="Error" />
+    <Rule Id="C6332" Action="Error" />
+    <Rule Id="C6333" Action="Error" />
+    <Rule Id="C6335" Action="Error" />
+    <Rule Id="C6381" Action="Error" />
+    <Rule Id="C6383" Action="Error" />
+    <Rule Id="C6384" Action="Error" />
+    <!-- warning C6385: invalid data: accessing buffer-name, the readable size is size1 bytes, but size2 bytes may be read: Lines: x, y -->
+    <Rule Id="C6385" Action="Warning" />
+    <!--warning C6386: buffer overrun: accessing <buffer name>, the writable size is <size1> bytes, but <size2> bytes may be written: Lines: x, y -->
```
Comment (dthaler) [inline Analyze.default.ruleset:243]:
```suggestion
    <!-- warning C6386: buffer overrun: accessing <buffer name>, the writable size is <size1> bytes, but <size2> bytes may be written: Lines: x, y -->
```

File: docs/DevelopmentGuide.md [hunk 4]
```diff
@@ -126,3 +126,13 @@ typedef struct _ebpf_widget
 Above all, if a file happens to differ in style from these guidelines (e.g.,
 private members are named `m_member` rather than `_member`), the existing style
 in that file takes precedence.
+
+### Project files
+
+All project files should include the following line to enable automated code analysis.
+
+```
+<Import Project="$(SolutionDir)\Analyze.default.props" />
```
Comment (dthaler) [inline docs/DevelopmentGuide.md:135]:
```suggestion
<Import Project="$(SolutionDir)Analyze.default.props" />
```
Actually I don't think this should be needed at all, and all the vcxproj changes can be reverted.
You can just use a Directory.Build.props file as documented in https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019

Comment (Alan-Jowett) [reply docs/DevelopmentGuide.md:135]: Thanks. I didn't know that. This is a much better option. :-)

**Review by dthaler — APPROVED** [review]

**Pull request merged**